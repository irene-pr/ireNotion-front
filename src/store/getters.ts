import router from "@/router";
import paths from "../router/paths";

const getters = {
  redirectToUserBoard(): void {
    router.push(paths.userBoard);
  },
  redirectToLogin(): void {
    router.push(paths.login);
  },
  redirectToRegister(): void {
    router.push(paths.register);
  },
  redirectToHome(): void {
    router.push(paths.home);
  },
};

export default getters;
