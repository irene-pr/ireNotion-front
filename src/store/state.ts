const state = {
  isLoggedIn: false,
  userContent: {
    name: "",
    boards: [],
  },
  userData: {
    userId: "",
    username: "",
  },
  themeHeaders: "night-mode",
  themeSurfaces: "day-mode",
  isBoardEditModal: false,
  isUpdateNoteModal: false,
  idForModal: "",
};

export default state;
