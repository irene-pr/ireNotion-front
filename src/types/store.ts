export interface IChecklist {
  isChecked: boolean;
  paragraph: string;
}

export interface INote {
  type?: string;
  color?: string;
  position?: number;
  title?: string;
  paragraph?: string;
  list?: Array<string | IChecklist>;
  id: string;
  userId?: string;
}

export interface IBoard {
  type: string;
  name: string;
  notes: Array<INote>;
  id: string;
}

export interface IUserContent {
  name: string;
  boards: Array<IBoard>;
}

export interface IUserData {
  userId: string;
  username: string;
}

export interface IState {
  isLoggedIn: boolean;
  userContent: IUserContent;
  userData: IUserData;
  isLoading: boolean;
  themeHeaders: string;
  themeSurfaces: string;
  isBoardEditModal: boolean;
  isUpdateNoteModal: boolean;
  isUpdateListNoteModal: boolean;
  isUpdateChecklistNoteModal: boolean;
  idForModal: string;
  listForModal: Array<string>;
}

export interface IThemeState {
  themeHeaders: string;
  themeSurfaces: string;
}

export interface IUserState {
  isLoggedIn: boolean;
  userContent: IUserContent;
  userData: IUserData;
  isLoading: boolean;
}

export interface IModalState {
  isBoardEditModal: boolean;
  isUpdateNoteModal: boolean;
  isUpdateListNoteModal: boolean;
  isUpdateChecklistNoteModal: boolean;
  idForModal: string;
  listForModal: Array<string>;
}
