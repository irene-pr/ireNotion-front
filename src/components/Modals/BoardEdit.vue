<template>
  <div
    class="modal-edit-board"
    :class="'modal-edit-board--' + themeSurfaces"
    @click="closeModal"
  >
    <div class="board-edit-modal" :class="'board-edit-modal--' + themeSurfaces" @click.stop>
      <font-awesome-icon
        icon="times"
        class="form__icon-close"
        @click="closeModal"
      ></font-awesome-icon>
      <h2 class="board-edit-modal__title">Edit the board name</h2>
      <form
        class="board-edit-form"
        @submit.prevent="onSubmit"
        autocomplete="off"
      >
        <div class="board-edit-form__name">
          <label class="board-edit-form__label" for="newBoardName">Name</label>
          <input
            class="board-edit-form__input"
            type="text"
            id="newBoardName"
            v-model="name"
            placeholder="New board name"
          />
        </div>
        <button
          class="board-edit-form__button"
          type="submit"
          :class="'board-edit-form__button--' + themeSurfaces"
          :disabled="name === ''"
        >
          Edit
        </button>
      </form>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { mapActions, mapState } from "vuex";
import { IBoard } from "@/types/interfaces";

export default defineComponent({
  name: "BoardEdit",
  data() {
    return {
      name: "",
    };
  },
  computed: {
    ...mapState(["userContent", "themeSurfaces", "idForModal"]),
  },
  methods: {
    ...mapActions(["editNameBoard", "setBoardEditModal", "setIdForModal"]),
    onSubmit() {
      this.editNameBoard({ idBoard: this.idForModal, newName: this.name });
      this.closeModal();
    },
    closeModal() {
      this.setBoardEditModal(false);
      this.setIdForModal("");
    },
  },
  mounted() {
    this.name = this.userContent.boards.filter(
      ({ id }: IBoard) => id === this.idForModal
    )[0].name;
  },
});
</script>

<style lang="scss">
@import "@/styles/_variables";
@import "@/styles/_extends";

.modal-edit-board {
  position: fixed;
  width: 100%;
  height: 100%;
  z-index: 20;
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: $Lato;

  &--day-mode {
    background-color: rgba(255, 255, 255, 0.5);

    input {
      color: $theme-dark-color;
    }
  }
  &--night-mode {
    background-color: rgba(0, 0, 0, 0.5);

    input {
      color: $theme-light-color;
    }
  }
}

.board-edit-modal {
  position: relative;
  width: 300px;
  height: 200px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  border-radius: 8px;

  &--day-mode {
    background-color: $theme-light-color;
    color: $theme-dark-color;
    border: 2px solid $theme-dark-color-opacity;
  }
  &--night-mode {
    background-color: $theme-dark-color;
    color: $theme-light-color;
    border: 2px solid $theme-light-color-opacity;
  }

  &__title {
    margin-bottom: 0;
    font-weight: 400;
  }
  .board-edit-form {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;

    &__name {
      margin: 25px 0;
    }
    &__label {
      padding-right: 20px;
    }
    &__input {
      height: 30px;
      padding-bottom: 0;
      border: none;
      border-bottom: 1px solid $theme-form-line;
      background-color: transparent;
      font-size: 14px;

      &:focus {
        outline: none;
        color: $theme-pink;
      }
    }

    &__button {
      @extend %button--little;
      margin-bottom: 10px;
      &--day-mode {
        background-color: rgba(255, 255, 255, 0.5);
        color: $theme-dark-color;
        border: 1px solid $theme-dark-color;
      }
      &--night-mode {
        background-color: rgba(0, 0, 0, 0.5);
        color: $theme-light-color;
        border: 1px solid $theme-light-color;
      }
    }
  }
  .form__icon-close {
    position: absolute;
    top: 10px;
    right: 10px;
  }
}
</style>
