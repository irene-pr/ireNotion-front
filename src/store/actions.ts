import { ActionContext } from "vuex";
import axios from "axios";
import jwtDecode from "jwt-decode";
import { IState, IUserLoginData } from "@/types/interfaces";
import router from "@/router";
import paths from "@/router/paths";

const actions = {
  async loginUser(
    { commit }: ActionContext<IState, IState>,
    user: IUserLoginData
  ): Promise<void> {
    const { data: token } = await axios.post(
      `${process.env.VUE_APP_API}/user/login/`,
      user
    );
    localStorage.setItem("token", JSON.stringify(token));
    commit("setUserData", jwtDecode(token.token));
  },

  async getUserContent({
    commit,
  }: ActionContext<IState, IState>): Promise<void> {
    const response = JSON.parse(localStorage.getItem("token") || "");
    const { data: userContent } = await axios.get(
      `${process.env.VUE_APP_API}/user/content/`,
      {
        headers: { Authorization: `Bearer ${response.token}` },
      }
    );
    if (response.status === 401) {
      router.push(paths.login);
    }
    commit("setUserContent", userContent);
  },

  async checkToken({
    commit,
  }: ActionContext<IState, IState>): Promise<string | void> {
    try {
      const token = JSON.parse(localStorage.getItem("token") || "");
      commit("setUserData", jwtDecode(token.token));
      return commit("setIsLoggedIn", true);
    } catch {
      return "User not logged in";
    }
  },

  async createParagraphNote(
    { dispatch }: ActionContext<IState, IState>,
    idBoard: string
  ): Promise<string | void> {
    const randomNumber = Math.floor(Math.random() * 4);
    const colors = ["yellow", "pink", "blue", "orange"];
    const initialParagraphCard = {
      type: "paragraph",
      title: "Title",
      paragraph: "This is a note. You can write whatever you want",
      color: colors[randomNumber],
    };
    const body = {
      note: initialParagraphCard,
      idBoard,
    };

    try {
      const { token } = JSON.parse(localStorage.getItem("token") || "");
      await axios.post(`${process.env.VUE_APP_API}/note/create/`, body, {
        headers: { Authorization: `Bearer ${token}` },
      });
      return dispatch("getUserContent");
    } catch {
      return "Paragraph note could not be created";
    }
  },
};

export default actions;
