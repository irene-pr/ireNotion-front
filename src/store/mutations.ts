import { IState, IUserContent } from "@/types/interfaces";

const mutations = {
  setToken(state: IState, payload: string): void {
    state.token = payload;
  },

  setUserContent(state: IState, payload: IUserContent): void {
    state.userContent = payload;
  },
};

export default mutations;
