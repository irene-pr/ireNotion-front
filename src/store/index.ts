import { createStore } from "vuex";
import actions from "./actions";
import mutations from "./mutations";
import state from "./state";
import getters from "./getters";

export default createStore({
  state,
  mutations,
  actions,
  getters,
  modules: {},
});
