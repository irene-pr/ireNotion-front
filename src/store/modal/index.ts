import state from "./state";
import mutations from "./mutations";
import actions from "./actions";

const modal = {
  namespaced: true,
  state,
  mutations,
  actions,
};

export default modal;
