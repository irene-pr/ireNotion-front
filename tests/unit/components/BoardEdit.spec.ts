import { mount } from "@vue/test-utils";
import { createStore } from "vuex";
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import { library } from "@fortawesome/fontawesome-svg-core";
import { fas } from "@fortawesome/free-solid-svg-icons";
import BoardEdit from "@/components/Modals/BoardEdit.vue";
import { IState } from "@/types/store";

library.add(fas);

describe("Given a BoardEdit component", () => {
  let mockedState: IState;
  let editNameBoardMock = jest.fn();
  let setBoardEditModalMock = jest.fn();
  let setIdForModalMock = jest.fn();
  let onSubmitMock = jest.fn();
  let closeModalMock = jest.fn();
  let store = createStore({});
  let options = {
    data() {
      return {
        name: "",
      };
    },
    components: {
      "font-awesome-icon": FontAwesomeIcon,
    },
    stubs: ["FontAwesomeIcon"],
    global: {
      plugins: [store],
    },
    mocks: {
      $store: store,
    },
  };
  function getOptions() {
    return {
      data() {
        return {
          name: "",
        };
      },
      components: {
        "font-awesome-icon": FontAwesomeIcon,
      },
      stubs: ["FontAwesomeIcon"],
      global: {
        plugins: [store],
      },
      mocks: {
        $store: store,
      },
    };
  }
  function getOptionsWithMethods() {
    return {
      data() {
        return {
          name: "",
        };
      },
      components: {
        "font-awesome-icon": FontAwesomeIcon,
      },
      stubs: ["FontAwesomeIcon"],
      global: {
        plugins: [store],
      },
      methods: {
        onSubmit: onSubmitMock,
        closeModal: closeModalMock,
      },
      mocks: {
        $store: store,
      },
    };
  }

  beforeEach(() => {
    mockedState = {
      isLoading: false,
      isLoggedIn: true,
      userContent: {
        name: "Franny",
        boards: [
          {
            type: "board",
            name: "tablero1",
            notes: [{ type: "paragraph", color: "yellow" }],
            id: "123",
          },
          {
            type: "board",
            name: "tablero2",
            notes: [{ type: "paragraph", color: "yellow" }],
            id: "234",
          },
        ],
      },
      userData: {
        userId: "",
        username: "",
      },
      themeHeaders: "night-mode",
      themeSurfaces: "day-mode",
      isBoardEditModal: true,
      isUpdateNoteModal: false,
      idForModal: "123",
    };
    editNameBoardMock = jest.fn();
    setBoardEditModalMock = jest.fn();
    setIdForModalMock = jest.fn();
    onSubmitMock = jest.fn();
    closeModalMock = jest.fn();
    store = createStore({
      state() {
        return mockedState;
      },
      actions: {
        editNameBoard: editNameBoardMock,
        setBoardEditModal: setBoardEditModalMock,
        setIdForModal: setIdForModalMock,
      },
    });
  });

  describe("When it is rendered", () => {
    test("Then it renders", () => {
      options = getOptions();
      mount(BoardEdit, options);
    });
    test("Then the it matches the snapshot", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);
      expect(wrapper.element).toMatchSnapshot();
    });
    test("Then name of the board is 'tablero1'", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);
      expect(wrapper.vm.$data.name).toBe("tablero1");
    });
    test("Then it renders a closing icon", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const closeIcon = wrapper.find(".form__icon-close");

      expect(closeIcon.exists()).toBeTruthy();
    });
    test("Then it renders a heading", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const heading = wrapper.find("h2");

      expect(heading.text()).toBe("Edit the board name");
    });
    test("Then it renders a form", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const form = wrapper.find("form");

      expect(form.exists()).toBeTruthy();
    });
  });
  describe("When the form is rendered", () => {
    test("Then a label for name is rendered", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const label = wrapper.find("label");

      expect(label.exists()).toBeTruthy();
    });
    test("Then an input for name is rendered", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const input = wrapper.find("input");

      expect(input.exists()).toBeTruthy();
    });
    test("Then an edit button is rendered", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const button = wrapper.find("button");

      expect(button.exists()).toBeTruthy();
    });
  });
  describe("When it renders a closing icon and you click it", () => {
    test("Then the method closeModal is called", () => {
      options = getOptionsWithMethods();
      const wrapper = mount(BoardEdit, options);

      const closeIcon = wrapper.find(".form__icon-close");
      closeIcon.trigger("click");

      expect(closeModalMock).toHaveBeenCalled();
    });
    test("Then the action setBoardEditModal is called", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const closeIcon = wrapper.find(".form__icon-close");
      closeIcon.trigger("click");

      expect(setBoardEditModalMock).toHaveBeenCalled();
    });
    test("Then the action setIdForModal is called", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const closeIcon = wrapper.find(".form__icon-close");
      closeIcon.trigger("click");

      expect(setIdForModalMock).toHaveBeenCalled();
    });
  });
  describe("When the modal is double clicked", () => {
    test("Then the method closeModal is called", () => {
      options = getOptionsWithMethods();
      const wrapper = mount(BoardEdit, options);

      const modal = wrapper.find(".modal-edit-board");
      modal.trigger("dblclick");

      expect(closeModalMock).toHaveBeenCalled();
    });
    test("Then the action setBoardEditModal is called", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const modal = wrapper.find(".modal-edit-board");
      modal.trigger("dblclick");

      expect(setBoardEditModalMock).toHaveBeenCalled();
    });
    test("Then the action setBoardEditModal is called", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const modal = wrapper.find(".modal-edit-board");
      modal.trigger("dblclick");

      expect(setIdForModalMock).toHaveBeenCalled();
    });
  });
  describe("When the input is left empty", () => {
    test("Then the edit button is disabled", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const input = wrapper.find("input");
      input.setValue("");

      const button = wrapper.find("button");

      expect(button.element.getAttributeNames()).toContain("disabled");
    });
  });
  describe("When input has been rewritten and the edit button clicked", () => {
    test("Then the onSubmit method gets called", () => {
      options = getOptionsWithMethods();
      const wrapper = mount(BoardEdit, options);

      const input = wrapper.find("input");
      input.setValue("Mario");

      const form = wrapper.find("form");
      form.trigger("submit.prevent");

      expect(onSubmitMock).toHaveBeenCalled();
    });
    test("Then the onSubmit method gets called", () => {
      options = getOptions();
      const wrapper = mount(BoardEdit, options);

      const input = wrapper.find("input");
      input.setValue("Mario");

      const form = wrapper.find("form");
      form.trigger("submit.prevent");

      expect(editNameBoardMock).toHaveBeenCalled();
    });
  });
});
