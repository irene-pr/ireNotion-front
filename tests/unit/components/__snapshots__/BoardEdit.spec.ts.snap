// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a BoardEdit component When it is rendered Then the it matches the snapshot 1`] = `
<div
  class="modal-edit-board modal-edit-board--day-mode"
>
  <div
    class="board-edit-modal board-edit-modal--day-mode"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-times fa-w-11 form__icon-close"
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
    <h2
      class="board-edit-modal__title"
    >
      Edit the board name
    </h2>
    <form
      autocomplete="off"
      class="board-edit-form"
    >
      <div
        class="board-edit-form__name"
      >
        <label
          class="board-edit-form__label"
          for="newBoardName"
        >
          Name
        </label>
        <input
          class="board-edit-form__input"
          id="newBoardName"
          placeholder="New board name"
          type="text"
        />
      </div>
      <button
        class="board-edit-form__button board-edit-form__button--day-mode"
        disabled=""
        type="submit"
      >
         Edit 
      </button>
    </form>
  </div>
</div>
`;
